app/jobs/autoscaling.py

```diff
@@ -260,227 +61,313 @@ def get_instance_tags():
-    aws_command(cmd)
-   
+                            _________________     Draining.
+                           |                 |    Scale-in protection still present.
+                           |    DRAINING     |    May still accept jobs for up to min_draining_wait seconds after entering this state.
+                           |_________________|    Can still finish running jobs for up to (min_draining_wait + job_tag_expiration) seconds after entering this state.
 
-def remove_termination_protection(instance_ids, asg):
-    if not instance_ids:
-        return
-    cmd = "aws autoscaling set-instance-protection --instance-ids {list_instances} --auto-scaling-group-name {asg_name} --no-protected-from-scale-in"
-    cmd = cmd.format(list_instances=' '.join(instance_ids), asg_name=asg['AutoScalingGroupName'])
-    aws_command(cmd)
 
+                                    |
+        remove scale-in protection  |
+                                   \ /
 
-def check_draining_servers(asg, tag_list, draining_tag, job_tag_prefix, job_tag_expiration, min_draining_wait, can_scale):
-    drain_date_by_instance_id = get_draining_servers(asg, tag_list, draining_tag)
-    num_jobs_by_instance_id = count_running_alignment_jobs(asg, tag_list, job_tag_prefix, job_tag_expiration)
-    instance_ids_to_terminate = []
-    current_timestamp = unixtime_now()
-    for instance_id, drain_date in drain_date_by_instance_id.items():
-        draining_interval = current_timestamp - drain_date
-        num_jobs = num_jobs_by_instance_id[instance_id]
-        if num_jobs == 0 and draining_interval > min_draining_wait:
-            instance_ids_to_terminate.append(instance_id)
-    if instance_ids_to_terminate:
-        print "The following instances are ready to be terminated: " + ", ".join(instance_ids_to_terminate)
-        if can_scale:
-            remove_termination_protection(instance_ids_to_terminate, asg)
-
-def count_running_batch_jobs():
-    queues_lock = threading.RLock()
-    queues = {"idseq_himem": [],
-              "idseq": []}
-    errors = []
-    def list_jobs(q):
-        try:
-            job_list = aws_command("aws batch list-jobs --job-queue {q} --job-status RUNNING".format(q=q))
-            with queues_lock:
-                queues[q] = json.loads(job_list)
-        except:
-            with queues_lock:
-                errors.append(q)
-    threads = [threading.Thread(target=list_jobs, args=[q]) for q in queues]
-    for t in threads:
-        t.start()
-    for t in threads:
-        t.join()
-    if errors:
-        return None
-    return sum(len(job_list["jobSummaryList"]) for job_list in queues.itervalues())
+                            _________________
+                           |                 |   Discarded.
+                           |   TERMINATING   |   Scale-in protection removed.
+                           |_________________|   Instance will terminate due to discrepancy between DesiredCapacity and actual number of instances.
```

Comment:
This is beautiful, the part I am curious about is "Instance will terminate due to discrepancy between DesiredCapacity and actual number of instances."    I think you mean, if desired capacity is lower than the current capacity and if it stays lower, then the instances here will be terminated.   But what if it doesn't stay lower -- if you come around in the next round of autoscaling and raise it.   What happens to any instances still in this state?

Reply:
The assumption (enforced in the rake code) is that incarnations of the autoscaler are spaced by enough time to prevent this case from occurring

Reply:
Update: put a check in place that should guarantee the case will not occur